### DISCUSSION TITLE
JustPy is awesome! / POST requests for webhook endpoints

### DISCUSSION THREAD

**Discussion by edwordout** [elimintz/justpy#635] [General]
Hey there beautiful people!

Firstly, I would like to say that I'm very grateful for being able to use JustPy to build my apps. Thank you, I really appreciate all  your work here!

About handling POST requests, I have not found it on your docs. I've read Starlette's docs and found a way as follows in the image:

![justpywebhook py](https://user-images.githubusercontent.com/94211097/204948378-130d479c-ba49-4cac-8dc2-d4fec0ce467c.png)

**Comment by Monallabs** [comment]
You will not be able to handle Post request with jp.SetRoute as it currently ignores the arguments pass to it (https://raw.githubusercontent.com/justpy-org/justpy/master/justpy/routing.py). 

I personally, don't use deocrators for routes. I would use app.add_route(path, func, methods=["GET"]) as documented here (https://github.com/encode/starlette/issues/675#issuecomment-541616651). 

I would convert the jpfunc (myFunc_POST in this case) to a starlette-response-returning function first
```
myfunc_response = app.response(myfunc_POST)
app.add_route(path, myfunc_response)
```

P.S.
@WolfgangFahl :  need to add kwargs to JustpyApp.jpRoute to handle this case.

**Reply by edwordout** [reply]
Hello Monallabs!

Looks great too. The code I posted works fine though.

![image](https://user-images.githubusercontent.com/94211097/205070970-7ac68f8c-c0e2-4106-9462-9f083a7bdbef.png)


Sorry for posting an image. It was my ever first post on github and I'm not really familiar with markdown syntax...
I've just figured it out how to properly add code here :). Try it out:

    import threading as tm
    from queue import Queue
    import justpy as jp
    
    def worker(q):
        while True:
            item = q.get()
            worker = tm.current_thread()
            print(f"\n-WORKER: {worker}")
            print(f"\n\t-STARTING: {item}\n")
            item.start()
            item.join()
            print(f"\n-WORKER: {worker}")
            print(f"\n\t-ENDED: {item}\n")
            q.task_done()
    
    webhook_worker_queue = Queue()
    tm.Thread(target=worker, name="webhook_worker", args=(webhook_worker_queue,), daemon=True).start()
    
    def body_rocking(request_body):
        from pprint import pp as pretty
        pretty(request_body)
        return
    
    @jp.SetRoute("/webhook/{endpoint}")
    async def myfunc_POST(request): # The _POST is important, keep reading...
        from json import dumps as jsondumps
    
        wp = jp.WebPage()
        wp.use_websockets = False
        wp.html = jsondumps({"status": "active"})
    
        endpoint = request.path_params["endpoint"]
    
        print("\n-NEW WEBHOOK CALL:\n")
        print(f"\t-ENDPOINT: {endpoint}\n")
    
        try:
            body = await request.json()
        except Exception as err:
            print(f"\n-!-BODY ERROR: {err}\n")
    
        t = tm.Thread(target=body_rocking, args=(body,), daemon=True)
        webhook_worker_queue.put(t)
    
        return wp
    
    app = jp.app
    
    REST_li = ["GET", "POST", "PUT", "PATCH", "DELETE"]
    
    print("\n- APP ROUTES:\n"); i = 1
    for route in app.router.routes:
        rtname = route.name
        if "_" in rtname:                   # Here I identify functions that
            nmethod = rtname.split("_")[1]  # were suffixed with a valid method
            if nmethod in REST_li:          # then I add the method to the methods set
                route.methods.add(nmethod)   # and we are golden!
        if getattr(route, "methods", []):
            print(f"\t{i} - {route}\n"); i += 1
    
    jp.justpy(None,
              host="0.0.0.0",
              port=8080)
